File: dsa/week(8)/stacktwo.cpp
#include <iostream>
using namespace std;
template <typename T>
class Node
{
public:
    T data;
    Node<T> *next;
    Node(T data)
    {
        this->data = data;
        next = nullptr;
    }
    ~Node()
    {
        delete next;
    }
};
template <typename T>
class Stack{
    Node<T> *head;
    Node<T> *tail;
    int size; // number of elements present in stack
public:
    Stack() {
        head = nullptr;
        tail = nullptr;
        size = 0;
    }
    int getSize(){
        return size;
    }
    bool isEmpty() // Check if the stack is empty
    {
        return head == nullptr;
    }
    void push(T element) // Insert the newNode at the end of the list and update the tail node
    {
        Node<T> *newnode = new Node<T>(element);
        if (isEmpty())
        {
            head = newnode;
            tail = newnode;
        }
        else
        {
            tail->next = newnode;
            tail = newnode;
        }
        size++;
    }
    T pop()
    {
        if (isEmpty())
        {
            cout << "Stack is empty. Cannot pop." << endl;
            return T(); // Return a default value for T
        }
        T data = tail->data;
        if (head == tail) // Only one element in the stack
        {
            delete tail;
            head = nullptr;
            tail = nullptr;
        }
        else
        {
            Node<T> *temp = head;
            while (temp->next != tail)
            {
                temp = temp->next;
            }
            delete tail;
            tail = temp;
            tail->next = nullptr;
        }
        size--;
        return data;
    }
    T top()
    {
        if (isEmpty())
        {
            cout << "Stack is empty. No top element." << endl;
            return T(); // Return a default value for T
        }
        return tail->data;
    }
    ~Stack()
    {
        while (!isEmpty())
        {
            pop();
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    Stack<int> stack;
    stack.push(10);
    stack.push(20);
    stack.push(30);

    cout << "Top element: " << stack.top() << endl;
    cout << "Stack size: " << stack.getSize() << endl;

    cout << "Popped element: " << stack.pop() << endl;
    cout << "Popped element: " << stack.pop() << endl;

    cout << "Top element: " << stack.top() << endl;
    cout << "Stack size: " << stack.getSize() << endl;

    return 0;
}

File: dsa/week(8)/stacktwo.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}
// Stack class definition
class Stack {
    constructor() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }
    getSize() {
        return this.size;
    }
    isEmpty() {
        return this.head === null;
    }
    push(element) {
        const newNode = new Node(element);
        if (this.isEmpty()) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            this.tail = newNode;
        }
        this.size++;
    }
    pop() {
        if (this.isEmpty()) {
            console.log("Stack is empty. Cannot pop.");
            return null; // Return null for empty stack
        }
        const data = this.tail.data;
        if (this.head === this.tail) { // Only one element in the stack
            this.head = null;
            this.tail = null;
        } else {
            let temp = this.head;
            while (temp.next !== this.tail) {
                temp = temp.next;
            }
            this.tail = temp;
            this.tail.next = null;
        }
        this.size--;
        return data;
    }
    top() {
        if (this.isEmpty()) {
            console.log("Stack is empty. No top element.");
            return null; // Return null for empty stack
        }
        return this.tail.data;
    }
}
// Main function to test the Stack class
function main() {
    const stack = new Stack();
    stack.push(10);
    stack.push(20);
    stack.push(30);

    console.log("Top element:", stack.top());
    console.log("Stack size:", stack.getSize());

    console.log("Popped element:", stack.pop());
    console.log("Popped element:", stack.pop());

    console.log("Top element:", stack.top());
    console.log("Stack size:", stack.getSize());
}


main();